pkg/reconciler/certificate/resources/cert_manager_certificate.go

```diff
@@ -30,6 +30,12 @@ func MakeCertManagerCertificate(cmConfig *config.CertManagerConfig, knCert *v1al
 	if len(knCert.Spec.DNSNames) > 0 {
 		commonName = knCert.Spec.DNSNames[0]
 	}
+	if len(commonName) > 63 {
+		// Max length of a CN for cert-manager is 64 characters:
+		// https://cert-manager.io/docs/reference/api-docs/#cert-manager.io/v1.CertificateSpec
+		commonName = commonName[:63]
```

Comment:
I am afraid this won't work at least for LetsEncrypt as the `commonName` needs to be a valid domain owned by the certificate requester otherwise the challenges cannot be passed and the certificate won't be issued.

I would suggest using the shortest `DNSNames` as the `commonName` which was also suggested at https://github.com/cert-manager/cert-manager/issues/1462#issuecomment-471502877. If the length of the shortest `DNSNames` is still longer than 63, we can return error.

WDYT?

Reply:
That doesn't seem to help, because our `kcert` resources seem to have one `dnsName`, like so:

```yaml
apiVersion: networking.internal.knative.dev/v1alpha1
kind: Certificate
metadata:
  annotations:
    networking.knative.dev/certificate.class: cert-manager.certificate.networking.knative.dev
    serving.knative.dev/creator: masterclient
    serving.knative.dev/lastModifier: masterclient
  labels:
    serving.knative.dev/route: test-123456789123456789123456789123456789123456789123456789
  name: route-619e26c6-665f-4304-a57f-4c0aad601c7f
  namespace: default
  ownerReferences:
  - apiVersion: serving.knative.dev/v1
    blockOwnerDeletion: true
    controller: true
    kind: Route
    name: test-123456789123456789123456789123456789123456789123456789
    uid: 619e26c6-665f-4304-a57f-4c0aad601c7f
  resourceVersion: "50589585"
spec:
  dnsNames:
  - test-123456789123456789123456789123456789123456789123456789.default.functions.majordemo.com
  secretName: route-619e26c6-665f-4304-a57f-4c0aad601c7f
status:
  conditions:
  - lastTransitionTime: "2022-03-24T22:47:11Z"
    message: Cert-Manager certificate has not yet been reconciled.
    reason: ReconcileFailed
    status: Unknown
    type: Ready
  observedGeneration: 3
```

This results in an error like the following:

```json
{
  "severity": "INFO",
  "timestamp": "2022-03-24T22:48:34.18727342Z",
  "logger": "net-certmanager-controller.certificate-controller.event-broadcaster",
  "caller": "record/event.go:282",
  "message": "Event(v1.ObjectReference{Kind:\"Certificate\", Namespace:\"default\", Name:\"route-619e26c6-665f-4304-a57f-4c0aad601c7f\", UID:\"c79039ce-5806-43c5-8310-c0246395d861\", APIVersion:\"networking.internal.knative.dev/v1alpha1\", ResourceVersion:\"50573288\", FieldPath:\"\"}): type: 'Warning' reason: 'CreationFailed' Failed to create Cert-Manager Certificate route-619e26c6-665f-4304-a57f-4c0aad601c7f/default: admission webhook \"webhook.cert-manager.io\" denied the request: spec.commonName: Too long: must have at most 64 bytes",
  "commit": "21ab5d8",
  "knative.dev/controller": "certificate-controller"
}
```

(So at least at this level, we report an error -- we might want to copy this up as a non-fatal error to the Route and Service.)

I was able to get the cert issued with the following patch:

```bash
kubectl patch kcert route-5efd61e6-686e-45fd-b1e8-4823a5c270c7 --type json --patch '[{"op":"add","path":"/spec/dnsNames/0","value":"a.knative-serving.functions.majordemo.com"}]'
```

i.e. add a short fake DNS name within our prefix (I used `a.knative-serving` because it seemed unlikely to collide, but we could also use something like `issue` if we want a shorter string).

Reply:
I think it is a good idea to see if we can provide a "fake" short `DNSName` as `CommonName`.

In your case, how did you get the certificate (by DNS challenge or HTTP01 challenge)?

I think for DNS challenge, this may work assuming that you own the domain `majordemo.com` and the cert-manager is able to configure the DNS server of `majordemo.com`.

I am not sure adding fake DNSName could work for HTTP01 challenge if the "fake" domain cannot be mapped to the ingress IP.

Reply:
Or probably we could make the "fake" short DNSName configurable by the cluster admin so that the cluster admin could make sure the "fake" short DNSName will mapped to the ingress IP when doing HTTP01 challenge.

Reply:
I used an HTTP01 challenge to get the certificate, not a DNS challenge.

I had a wildcard cert set up for `functions.majordemo.com`, and used that as my domain name for all namespaces in `config-domain`.

Reply:
It looks like we don't currently provide any information in the `kcert` resource to understand what the top-level domain used by the Route is.

Would you prefer to add this property as a `spec` field or an annotation? I'm leaning towards `spec`, but don't feel strongly.

Reply:
Do you suggest that we extract the top level domain, and configure `kcert` to obtain one TLS certificate for both top level domain (`functions.majordemo.com` in your case) and the service specific domain (the domain longer than 63 char in your case) ?

I think this is the good path forward. But I am not sure if we want to get certificates for top level domain (implicitly) or explicitly ask cluster admin to provide a "fake" domain just for bypassing this DNSName issue.  For the first option, I am afraid that there could be a security concern if certs could cover the top level domain, and in the meantime the top level domain is actually used by the user because if any cert is leaked, it could affect the top level domain. WDYT?

Reply:
As an end user, I'd like to add my perspective for how I would expect this to work. Instead of automatically using top-level domain (which as mentioned before could lead to unexpected security issues), KNative could expose a `common-name-template` config. It could default to `""` which falls back to the current behavior, otherwise the example `config-network` could recommend `"{{.Domain}}"` or `"placeholder.{{.Domain}}"`. Then, there should be a clear section in the documentation about CNs longer than 63 chars explaining the nuance of using the top-level domain vs. some other placeholder and what the requirements should be in each case.